rask-engine/src/resources/library.rs

```diff
@@ -0,0 +1,73 @@
+use super::Resource;
+use crate::EngineError;
+
+/// The library is used to store and retrieve resources.
+pub struct ResourceTable(&'static mut [Resource]);
+
+macro_rules! get_store {
+    ($type: ty, $enum_type: ident) => {
+        impl GetStore<$type> for ResourceTable {
+            unsafe fn get(&'static self, id: usize) -> Result<&'static $type, EngineError> {
+                self.index_check(id)?;
+                match &self.0[id] {
```

Comment:
Instead of using `Index`, you could use `get` (and `get_mut` in `store`) for indexing (`self.0.get(id)`), which returns an `Option`. That would be much cleaner than the `index_check`-function.

Reply:
I disagree if the requested resource is out of bounds, an error should be raised, because it is unexpected behaviour. It is essentially  a fancy panic.
And I don't like the approach of get_mut(), because with the current implementation we do all the wrapping in the library which would then needed to be done by the user.

Reply:
Panic? I thought a panic could not occur in your function. I don't get your point. You could do something like
```rust
match self.0.get(id).ok_or(myerrorblahblah)? {
}
```
This does not make any difference for the user

Reply:
It is also relevant, whether the error occured because of an out of bounds, or because of a resource type mismatch

Reply:
Of course, but that does not change anything on my point. See, you could do the following:
```rust
unsafe fn get(&'a self, id: usize) -> Result<&'a $type, EngineError> {
    match self.0.get(id) {
        Some(Resource::$enum_type(value)) => Ok(&value),
        Some(_) => Err("Wrong resource type".into()),
        None => Err("Invalid resource id")
    }
}
```